assets/js/editor-components/incompatible-payment-gateways-notice/use-incompatible-payment-gateways-notice.ts

```diff
@@ -0,0 +1,95 @@
+/**
+ * External dependencies
+ */
+import { useSelect } from '@wordpress/data';
+import { useEffect, useState } from '@wordpress/element';
+import { useLocalStorageState } from '@woocommerce/base-hooks';
+
+/**
+ * Internal dependencies
+ */
+import { STORE_KEY as PAYMENT_STORE_KEY } from '../../data/payment/constants';
+
+const initialDismissedNotices: React.SetStateAction< object[] > = [];
+
+const areEqual = ( array1: string[], array2: string[] ) => {
+	if ( array1.length !== array2.length ) {
+		return false;
+	}
+
+	const uniqueCollectionValues = new Set( [ ...array1, ...array2 ] );
+
+	return uniqueCollectionValues.size === array1.length;
+};
+
+export const useIncompatiblePaymentGatewaysNotice = (
+	blockName: string
+): [ boolean, () => void, { [ k: string ]: string }, number ] => {
+	const [ dismissedNotices, setDismissedNotices ] = useLocalStorageState(
+		`wc-blocks_dismissed_incompatible_payment_gateways_notices`,
+		initialDismissedNotices
+	);
+	const [ isVisible, setIsVisible ] = useState( false );
+
+	const { incompatiblePaymentMethods } = useSelect( ( select ) => {
+		const { getIncompatiblePaymentMethods } = select( PAYMENT_STORE_KEY );
+		return {
+			incompatiblePaymentMethods: getIncompatiblePaymentMethods(),
+		};
+	}, [] );
+	const incompatiblePaymentMethodsIDs = Object.keys(
+		incompatiblePaymentMethods
+	);
+	const numberOfIncompatiblePaymentMethods =
+		incompatiblePaymentMethodsIDs.length;
+
+	const isDismissedNoticeUpToDate = dismissedNotices.some(
+		( notice ) =>
+			Object.keys( notice ).includes( blockName ) &&
+			areEqual(
+				notice[ blockName as keyof object ],
+				incompatiblePaymentMethodsIDs
+			)
+	);
+
+	const isDismissed =
+		numberOfIncompatiblePaymentMethods === 0 || isDismissedNoticeUpToDate;
+	const dismissNotice = () => {
+		const dismissedNoticesSet = new Set( dismissedNotices );
+		dismissedNoticesSet.add( {
+			[ blockName ]: incompatiblePaymentMethodsIDs,
+		} );
+		setDismissedNotices( [ ...dismissedNoticesSet ] );
+	};
+
+	// This ensures the modal is not loaded on first render. This is required so
+	// Gutenberg doesn't steal the focus from the Guide and focuses the block.
+	useEffect( () => {
+		setIsVisible( ! isDismissed );
+
+		if ( ! isDismissed && ! isDismissedNoticeUpToDate ) {
+			setDismissedNotices( ( previousDismissedNotices ) =>
+				previousDismissedNotices.reduce( ( acc, curr ) => {
+					if ( Object.keys( curr ).includes( blockName ) ) {
+						return acc;
+					}
+					acc.push( curr );
+
+					return acc;
+				}, [] )
```

Comment:
We have a TS error for this state setter! @opr, what are your thoughts on it? It looks like doing a functional update isn't recognized by TS in this instance!

Reply:
Ah yes I see that. OK I would recommend the following:

- Create a type: `type StoredIncompatibleGateway = {
	[ k: string ]: string[];
};`
- Change `initialDismissedNotices` to `const initialDismissedNotices: React.SetStateAction< StoredIncompatibleGateway[] > = [];`
- Change the generic type passed to `useLocalStorageState` to `StoredIncompatibleGateway[]`
- Update `useLocalStorageState` to return `[ T, Dispatch< SetStateAction< T > > ]` (`Dispatch` and `SetStateAction` are types from `React`)

There will still be a couple of errors that I didn't look into fixing in detail, but they shouldn't be too tough to figure out!

Let me know if you'd like more help with that.

Reply:
Thank you! This worked perfectly!